pkg/webrtcclient/client.go

```diff
@@ -0,0 +1,234 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package webrtcclient
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/pion/webrtc/v3"
+)
+
+type Signaling struct {
+	SendCh chan interface{}
+	RecvCh chan map[string]interface{}
+	// The ICE servers to use in the webRTC connection.
+	ICEServers []webrtc.ICEServer
+	// The servers that were created client side and need to be sent to the device.
+	// This is typically a subset of Servers. Ignored if empty.
+	ClientICEServers []webrtc.ICEServer
+}
+
+type Controller struct {
+	signaling      *Signaling
+	peerConnection *webrtc.PeerConnection
+	observer       Observer
+}
+
+func (c Controller) connect(onComplete func(error)) {
+	// Provide sufficient buffer in case several events are triggered when the peer
+	// connection fails or closes.
+	closeCh := make(chan bool, 10)
+	c.peerConnection.OnConnectionStateChange(func(s webrtc.PeerConnectionState) {
+		switch s {
+		case webrtc.PeerConnectionStateConnected:
+			// Connected successfully
+			onComplete(nil)
+		case webrtc.PeerConnectionStateFailed:
+			onComplete(fmt.Errorf("Peer Connection failed"))
+			closeCh <- true
+		case webrtc.PeerConnectionStateClosed:
+			onComplete(fmt.Errorf("Peer Connection closed unexpectedly"))
+			closeCh <- true
+		default:
+			// These are temporary states, ignore.
+		}
+	})
+
+	c.peerConnection.OnICECandidate(func(candidate *webrtc.ICECandidate) {
+		if candidate == nil {
+			return
+		}
+		c.sendICECandidate(candidate.ToJSON())
+	})
+
+	// Handle signaling messages in a new go routine so this one can return when
+	// the connection is established.
+	go func() {
+		err := c.recvLoop(closeCh)
+		if err != nil {
+			c.observer.OnError(err)
+		}
+	}()
+
+	// Request offer from the device to start the signaling process.
+	c.signaling.SendCh <- NewRequestOfferMsg(c.signaling.ClientICEServers)
+}
+
+func (c Controller) recvLoop(closeCh chan bool) error {
+	for {
+		select {
+		case _ = <-closeCh:
+			return nil
+		case msg, open := <-c.signaling.RecvCh:
+			if !open {
+				return fmt.Errorf("Signaling channel closed")
+			}
+			switch msg["type"] {
+			case RequestOfferMsgType:
+				panic("Device requested an offer. This violates the signaling protocol")
+			case OfferMsgType:
+				var offer webrtc.SessionDescription
+				err := Reshape(msg, &offer)
+				if err != nil {
+					// msg was obtained from a JSON string so Reshape shouldn't fail
+					panic(fmt.Errorf("Failed to reshape json: ", err))
```

Comment:
nit: if this is just for printing the panic message why not use fmt.Sprintf instead of creating a new error.